### DISCUSSION TITLE
Resistance based an applied voltage

### DISCUSSION THREAD

**Discussion by termostatos** [SpiceSharp/SpiceSharp#129] [Q&A] [answered]
Hello,
I wonder if there is a way to assign a resistance value depending on the voltage. 

I attach the de Spice's scheme that we want to replicate in Spice Sharp and also the Spice NetList.

Sorry for the inconvenience or if doubts cannot be resolved here, if that is the case and there is some way to raise them I would like to know how.

Thank you for your attention to this matter.

Spice Scheme:
![imagen](https://user-images.githubusercontent.com/33329368/102062574-39be8080-3df5-11eb-80ba-8ab1e8483356.png)

SpiceNetlist:
V1 N002 N005 750 Rser=0.2
R1 uno N002 0.024
R2 tres N005 0.015
R3 dos uno 0.12
R4 cuatro tres 0.095
R5 N003 dos 0.23
R6 N006 cuatro 0.18
R7 0 N005 0.8
R8 0 N006 0.8
V2 N003 N006 750 Rser=0.2
R9 uno tres R=V(uno)**2/250000
R10 dos cuatro R=V(dos)**2/30000
V3 N001 N004 750 Rser=0.2
R11 cinco N001 0.024
R12 seis N004 0.015
R13 siete cinco 0.12
R14 ocho seis 0.095
R15 N002 siete 0.23
R16 N005 ocho 0.18
R17 0 N004 0.8
R19 cinco seis R=V(cinco)**2/750000
R20 siete ocho R=V(siete)**2/124000
.op
.backanno
.end

**Comment by svenboulanger** [comment, marked as answer]
Hi @termostatos!

Thanks for taking a look at Spice#!

What you have described is actually part of [Spice#.Behavioral](https://github.com/SpiceSharp/SpiceSharpBehavioral). You can use that package to create a behavioral voltage source and specify - as your own custom function - how the current depends on its own current to model a variable resistor.

In your case, the line `R19 cinco seis R=V(cinco)**2/750000` would become a behavioral voltage source between `cinco` and `ceis` with the expression `V(behavioral_voltage_source)*V(cinco)^2/750000` (i.e. V=I*R).

The unit test and perhaps example where this kind of stuff is tested is [here for the behavioral voltage source](https://github.com/SpiceSharp/SpiceSharpBehavioral/blob/13c63c197439d15c086fdae940b1a5f100bb0c77/SpiceSharpBehavioralTest/Components/BehavioralVoltageSourceTests.cs#L50), or if you rather express it as a conductance you can also use its counterpart [here for the behavioral current source](https://github.com/SpiceSharp/SpiceSharpBehavioral/blob/13c63c197439d15c086fdae940b1a5f100bb0c77/SpiceSharpBehavioralTest/Components/BehavioralCurrentSourceTests.cs#L52).

**Reply by svenboulanger** [reply]
I just noticed that Spice#.Behavioral is not up to date with v3.0.3 yet. This is being updated as I write this.

**Reply by termostatos** [reply]
Hi Sven: 
Thank you very much for the quick reply and the given information.
Yesterday I was doing some comparisons between our model in LTSpice and your propousal and din´t fit the results that we obtain. 
Cheking my code I think I have some mistakes naming the behavioral voltage but now I can't check it because it throws an excepction in SpiceSharpBehavioral library which name is "Invalid function call" even in the example given in [Spice#.Behavioral](https://github.com/SpiceSharp/SpiceSharpBehavioral).

**Reply by svenboulanger** [reply]
Hi, can you share some code in the question? That exception is thrown when it cannot parse function arguments correctly.

**Reply by termostatos** [reply]
Yes, I create a Console proyect with Visual Studio 2019 and try to run the example

```
using System;
using SpiceSharp;
using SpiceSharp.Components;
using SpiceSharp.Simulations;

namespace ConsoleSpiceBejaviour
{
    class Program
    {
        static void Main(string[] args)
        {
            var ckt = new Circuit(
                new VoltageSource("V1", "in", "0", 0.0),
                new BehavioralVoltageSource("A1", "out", "0", "V(in)^2+2"));

            var dc = new DC("dc", "V1", -1, 1, 0.1);
            dc.ExportSimulationData += (sender, e) =>
            {
                Console.WriteLine(e.GetVoltage("out"));
            };
            dc.Run(ckt);
            Console.ReadKey();
        }
    }
}
```

**Reply by svenboulanger** [reply]
Hi, by default functions are case-sensitive. If you use `v(in)^2+2` it will work. Maybe I should make it case-insensitive by default (same as regular Spice)...

**Reply by termostatos** [reply]
Thank you very much!! It works now.

**Reply by svenboulanger** [reply]
Can I mark this as answered?

**Reply by termostatos** [reply]
Sure!!
